src/main/java/org/truffleruby/core/numeric/BignumNodes.java

```diff
@@ -700,20 +701,37 @@ public Object compareFallback(DynamicObject a, DynamicObject b) {
     @Primitive(name = "bignum_pow")
     public static abstract class BignumPowPrimitiveNode extends PrimitiveArrayArgumentsNode {
 
-        private final ConditionProfile negativeProfile = ConditionProfile.createBinaryProfile();
+        private static final int BIGLEN_LIMIT = 32 * 1024 * 1024;
+
+        public static BignumPowPrimitiveNode create() {
+            return BignumNodesFactory.BignumPowPrimitiveNodeFactory.create(null);
+        }
+
+        public abstract Object executePow(Object a, Object b);
 
         @Specialization
         public Object pow(DynamicObject a, int b) {
-            return pow(a, (long) b);
+            return executePow(a, (long) b);
         }
 
         @Specialization
-        public Object pow(DynamicObject a, long b) {
+        public Object pow(DynamicObject a, long b,
+                @Cached("createBinaryProfile()") ConditionProfile negativeProfile,
+                @Cached("createBinaryProfile()") ConditionProfile maybeTooBigProfile,
+                @Cached("new()") WarnNode warnNode) {
             if (negativeProfile.profile(b < 0)) {
                 return FAILURE;
             } else {
+                final BigInteger value = Layouts.BIGNUM.getValue(a);
+                final int xbits = value.bitLength();
+
+                if (maybeTooBigProfile.profile(xbits > BIGLEN_LIMIT || (xbits * b > BIGLEN_LIMIT))) {
```

Comment:
Is this from MRI?
Isn't `xbits * b` always bigger, unless `b` is 0?

Reply:
It's from MRI. `BIGLEN_LIMIT` is fairly large (33,554,432) and `xbits` is likely fairy small. That leaves a lot of room for `b`.

In the case of `2 ** 427387904`, by the time `Fixnum#**` overflows to `Bignum#**`, the `Bignum` value has an `xbits` value of 65.

Reply:
So this condition is taken from MRI?
Could you just add a comment so we don't wonder?

Reply:
Will do.